django/db/models/sql/query.py

```diff
@@ -65,12 +65,33 @@ def get_field_names_from_opts(opts):
     )
 
 
+def get_paths_from_expression(expr):
+    if isinstance(expr, F):
+        yield expr.name
+        return
+
+    children = [
+        child_expr for child_expr in expr.flatten() if isinstance(child_expr, (Q, F))
+    ]
+    for child_expr in children:
+        if isinstance(child_expr, F):
+            yield child_expr.name
+        elif isinstance(child_expr, Q):
+            yield from get_children_from_q(child_expr)
```

Comment:
Do we need a temporary list? Maybe:
```suggestion
    if isinstance(expr, F):
        yield expr.name
    elif hasattr(expr, "flatten"):
        for child in expr.flatten():
            if isinstance(child, F):
                yield child_expr.name
            elif isinstance(child, Q):
                yield from get_children_from_q(child_expr)
```

or even
```python
def get_paths_from_expression(expr):
    if isinstance(expr, F):
        yield expr.name
    elif hasattr(expr, "flatten"):
        for child in expr.flatten():
            if isinstance(child, Q):
                yield from get_children_from_q(child)
            else:
                yield from get_paths_from_expression(child)
```